### DISCUSSION TITLE
How to use Harfbuzz for vertical text

### DISCUSSION THREAD

**Discussion by my2iu** [harfbuzz/harfbuzz#3294] [Q&A] [answered]
I'm trying to bumble my way through understanding opentype, Harfbuzz, and vertical text. I just want to check if I'm doing things correctly, but I'm not sure where else to ask.

Does this procedure for figuring out glyph positions for vertical text using Harfbuzz seem correct?

1. Get a run of vertical text and feed it into Harfbuzz to get the glyphs and glyph positions
2. Take the resulting glyph positions and adjust them using the values from hb_font_subtract_glyph_origin_for_direction
3. The top-side bearing from the vmtx table should be subtracted from the y-position of each glyph
4. The maximum y extent of each glyph should then be subtracted from the y-position of each glyph
5. The x position should be further adjusted by vertTypoDescender from the vmtx table if I want a central baseline
6. Draw the glyphs at those positions

**Comment by behdad** [comment, marked as answer]
Steps 2 to 5 should not be necessary. Tell us more about your setup.

**Reply by my2iu** [reply]
I'm using Harfbuzz from JavaScript, so I'm just compiling Harfbuzz using HB_TINY to WebAssembly and calling into it to get glyphs and glyph positions. I then use a custom version of Typr.js to extract the glyph outlines, and I have some of my own code for rendering to a SDF and then displaying it using WebGL. I pretty much just draw the glyph vectors exactly at the glyph positions returned by Harfbuzz.

I don't think the JavaScript is causing problems. The positions I get seem to be consistent with me just calling Harfbuzz in C++ directly. So if I compile Harfbuzz with HB_TINY and then try to get some vertical glyph positions like this:

```
hb_blob_t *blob = hb_blob_create_from_file_or_fail ("../../NotoSansSC-Regular.otf");
hb_buffer_t *buf = hb_buffer_create();
hb_codepoint_t codepoints[] = {0x5165, 0x4e00, 0x5165, 0x5165};
hb_buffer_add_codepoints(buf, codepoints, 4, 0, 4);
hb_buffer_set_direction(buf, HB_DIRECTION_TTB);
hb_buffer_set_script(buf, HB_SCRIPT_HAN);
hb_buffer_set_language(buf, hb_language_from_string("zh", -1));
hb_shape(font, buf, NULL, 0);
```

I get back

```
0: glyph 9568  x-advance: 0  x-shift: -500  y-advance: -1000  y-shift: -1160
1: glyph 8699  x-advance: 0  x-shift: -500  y-advance: -1000  y-shift: -1160
2: glyph 9568  x-advance: 0  x-shift: -500  y-advance: -1000  y-shift: -1160
3: glyph 9568  x-advance: 0  x-shift: -500  y-advance: -1000  y-shift: -1160
```

So the character positions are too low. The y-shift is lower than the y-advance, so part of each character descends below the caret. I noticed in #573 and #2765 that people recommended using add/subtract__glyph_origin_for_direction, so that's why I did step 2. When I do that, I get back

```
0: glyph 9568  x-advance: 0  x-shift: 0  y-advance: -1000  y-shift: 0
1: glyph 8699  x-advance: 0  x-shift: 0  y-advance: -1000  y-shift: 0
2: glyph 9568  x-advance: 0  x-shift: 0  y-advance: -1000  y-shift: 0
3: glyph 9568  x-advance: 0  x-shift: 0  y-advance: -1000  y-shift: 0
```

That's just the upper-left corner of each glyph, so I did step 3-5 to center each glyph in its character box/frame as described in the [opentype docs](https://docs.microsoft.com/en-us/typography/opentype/spec/vmtx).

But I'm not sure if I'm following the correct procedure for using Harfbuzz. NotoSans is a CFF font. Could that be causing problems with HB_TINY?

**Comment by behdad** [comment]
The result you are getting looks almost correct except for what you point out. Can you attach a standalone C++ code showing this, so I can debug?

**Reply by my2iu** [reply]
Here's my steps for some standalone C++ code:

1. I download harfbuzz 3.1.1 (though I originally tried getting vertical text working back with 2.7.4 and had the same problems then)
2. I go into the src directory
3. I make a `vertical.cc` file with the following code: 
``` 
#include "hb.h"
#include <cstdio>

int main(void)
{
	hb_blob_t *blob = hb_blob_create_from_file_or_fail ("NotoSansSC-Regular.otf");
	
	hb_face_t* face = hb_face_create(blob, 0);
	hb_font_t* font = hb_font_create(face);
	
	hb_buffer_t *buf = hb_buffer_create();
	hb_codepoint_t codepoints[] = {0x5165, 0x4e00, 0x5165, 0x5165};
	hb_buffer_add_codepoints(buf, codepoints, 4, 0, 4);
	hb_buffer_set_direction(buf, HB_DIRECTION_TTB);
	hb_buffer_set_script(buf, HB_SCRIPT_HAN);
	hb_buffer_set_language(buf, hb_language_from_string("zh", -1));

	hb_shape(font, buf, NULL, 0);
	unsigned int glyph_count;
      	hb_glyph_info_t *glyph_info    = hb_buffer_get_glyph_infos(buf, &glyph_count);
      	hb_glyph_position_t *glyph_pos = hb_buffer_get_glyph_positions(buf, &glyph_count);
	for (int n = 0; n < glyph_count; n++)
	{
		printf("%d: %d x: %d %d y: %d %d\n", n, glyph_info[n].codepoint, 
			glyph_pos[n].x_advance, glyph_pos[n].x_offset, 
			glyph_pos[n].y_advance, glyph_pos[n].y_offset);
	}
	
	// Adjust glyph origins
	for (int n = 0; n < glyph_count; n++)
	{
		hb_position_t x = 0, y = 0;
        	hb_font_add_glyph_origin_for_direction( font,
                            glyph_info[n].codepoint,
                            HB_DIRECTION_TTB,
                            &x ,
                            &y );
		printf("%d: %d x: %d %d y: %d %d\n", n, glyph_info[n].codepoint, 
			glyph_pos[n].x_advance, glyph_pos[n].x_offset + x, 
			glyph_pos[n].y_advance, glyph_pos[n].y_offset + y);
	}

	hb_buffer_destroy(buf);
	return 0;
}
```
4. I grab a copy of NotoSans for simplified Chinese and put it there 
5. I modify `hb_config.hh` to not define HB_NO_OPEN even if HB_TINY is defined
6. I compile with `g++ -DHB_TINY harfbuzz.cc vertical.cc`

**Comment by behdad** [comment]
I can reproduce the issue with your code. With hb-shape however I get -880 instead of -1160. Investigating.
```
$ ../build/util/hb-shape ./NotoSansSC-Regular.otf -u '0x5165, 0x4e00, 0x5165, 0x5165' --direction t --language zh --font-funcs=ot
[gid9629=0@-500,-880+0,-1000|gid8760=1@-500,-880+0,-1000|gid9629=2@-500,-880+0,-1000|gid9629=3@-500,-880+0,-1000]
```

**Reply by my2iu** [reply]
Oh, if it’s just a configuration issue, then I can probably debug it my end. I just couldn’t figure out if I was using the API correctly or not. Thanks!

**Reply by behdad** [reply]
Your code looks correct. So I like to debug it.

**Reply by my2iu** [reply]
It looks like HB_TINY defines HB_NO_CFF. And it looks like that causes the VORG and cff tables to not load. The NotoSans CJK fonts are CFF fonts, and without those tables, I guess Harfbuzz can't vertically layout those fonts properly. When I comment out the HB_NO_CFF part, I do get the expected -880 values.

I only defined HB_TINY because that's what the build script for harfbuzzjs used. Is it possible to make an alternative preprocessor macro that can make a small harfbuzz that also has CFF support? I think the Noto fonts are the main fonts that people are including in projects for internationalization now because of their completeness and licensing, so I think other people might need an easy configuration option for that combination too.

**Reply by behdad** [reply]
Should be fixed with HB_TINY. Please test.

**Reply by my2iu** [reply]
I'm starting to test now, but I'm still wondering if there should be a separate option for reenabling CFF with HB_TINY. It's true that the Noto fonts have a VORG table, but the Microsoft spec says that the VORG table is optional and that implementations should be able to create their own VORG table by parsing the CFF tables and drawing the glyphs themselves. On the other hand, I have no idea whether CFF fonts without VORG tables actually exist. 

Alternately, since consumers of Harfbuzz will have their own glyph rendering code that parses CFF, maybe Harfbuzz could somehow indicate that there was no VORG or glyf data, and that the renderer should adjust the glyph positions with the CFF data themselves. Or maybe the renderer would know there's no VORG or glyf data, and can just rely on the fact that Harfbuzz will act predictably so that it can add in the CFF max-y-extent data itself.

**Reply by my2iu** [reply]
Yes, Harfbuzz when compiled to WASM using HB_TINY seems to give correct y-shift values for NotoSans CJK fonts:
![hohoho](https://user-images.githubusercontent.com/1350963/145305292-ce68c9b6-3109-4850-9cd4-8a9950d91dec.png)
Thanks!

**Reply by behdad** [reply]
VORG is required. Finding glyph extents from CFF outlines is *REALLY* expensive. So we don't do it.

If VORG is missing, we use the horizontal ascent as glyph vertical origin. That's where the 1160 comes from. Unusual value, yes.

**Reply by my2iu** [reply]
That makes sense. The actual glyph renderer has to do that work anyway, so there’s no point in Harfbuzz doing it twice. I suppose, on the off-chance that such a font is encountered, the renderer can subtract out the horizontal ascent and use vmtx and its own calculated glyph extent information to get a better looking result.

**Reply by my2iu** [reply]
I’m being off-topic a bit, but do you know why the vmtx table doesn’t just give the position of the origin directly? Why does the spec need to separate the vertical shift data into two parts like that? I guess maybe they want the vertical layout code to exactly mirror the horizontal layout code, but it just seems unnecessarily complicated to me.

**Reply by behdad** [reply]
> I’m being off-topic a bit, but do you know why the vmtx table doesn’t just give the position of the origin directly? Why does the spec need to separate the vertical shift data into two parts like that? I guess maybe they want the vertical layout code to exactly mirror the horizontal layout code, but it just seems unnecessarily complicated to me.

My guess is for it to work with bitmap images more efficiently, such that bitmaps can be tight... glyf table encodes lsb/tsb directly so the glyph does not need to be rasterized to know lsb/tsb. CFF does not encode such data, so VORG was added.

**Reply by my2iu** [reply]
Oh, right. That must be from back when they had the philosophy that it’s more important for things to look good on a screen than to match the metrics when printed out or displayed on a device with a different dpi.
